### PR TITLE
feat(humming): FP8 DeepEP dispatch for humming MoE backend

### PR DESCRIPTION
**Pull request by guzekai01**
## Summary

FP8 DeepEP dispatch for the Humming MoE backend: adds `moe_permute_with_scale`
(row-permutes hidden states together with their group-128 FP8 scales), dispatch
input validation, and a fused masked activation+quant fast path. Bumps
`humming-kernels` to 0.1.12.

## How to run

Server (2 nodes × 8 H20, DeepSeek-V4-Pro pp2/tp8/ep8; run on each node with
`--node-rank 0/1`):

```bash
# fp8 arm only:
export SGLANG_HUMMING_INPUT_QUANT_CONFIG='{"dtype": "float8e4m3", "group_size": 128}'

python3 -m sglang.launch_server \
  --model-path DeepSeek-V4-Pro --trust-remote-code \
  --tp-size 8 --ep-size 8 --pp-size 2 \
  --nnodes 2 --node-rank <0|1> --dist-init-addr <rank0-ip>:23456 \
  --moe-runner-backend humming --moe-a2a-backend deepep \
  --deepep-dispatcher-output-dtype fp8 \
  --chunked-prefill-size 8192 --mem-fraction-static 0.85 \
  --disable-radix-cache --cuda-graph-max-bs-decode 64
```

(baseline arm: `--deepep-dispatcher-output-dtype bf16`, no quant-config env)

Accuracy client:

```bash
sgl-eval run gsm8k --base-url http://127.0.0.1:30000/v1 \
  --num-examples 1319 --num-threads 256 --temperature 0
```

Perf client (per cell):

```bash
python3 -m sglang.bench_serving --backend sglang --dataset-name random \
  --random-input-len <ISL> --random-output-len <OSL> \
  --num-prompts <N> --max-concurrency <C> \
  --random-range-ratio 1.0 --seed 1024 --request-rate inf \
  --flush-cache --warmup-requests 1
```

## Results (H20 ×16)

GSM8K (full 1319, greedy):

| dispatch | score | empty outputs |
|---|---|---|
| fp8 | **96.13%** | 0 |
| bf16 | 95.91% | 0 |

Perf (fp8 vs bf16, same weights):

| cell | fp8 | bf16 | Δ |
|---|---|---|---|
| TTFT 16k (mean) | 1386 ms | 1794 ms | **-22.8%** |
| TTFT 64k (mean) | 4395 ms | 5749 ms | **-23.6%** |
| TPOT 1k bs1 (median) | 29.7 ms | 46.3 ms | **-35.8%** |
| TPOT 64k bs4 (median) | 40.8 ms | 62.9 ms | **-35.1%** |

---
### CI States

Latest PR Test (Base): :x: [Run #32646675353](https://github.com/sgl-project/sglang/actions/runs/32646675353)
Latest PR Test (Extra): :x: [Run #32646675225](https://github.com/sgl-project/sglang/actions/runs/32646675225)
Latest PR Test (AMD ROCm 7.2): :x: [Run #32646675339](https://github.com/sgl-project/sglang/actions/runs/32646675339)

### CODE REVIEW

File: test/manual/layers/moe/test_humming_fp8_dispatch.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
+import pytest
+import torch
+
+from sglang.jit_kernel.dsv4.moe import silu_and_mul_masked_post_quant
+from sglang.srt.layers.moe.ep_moe.kernels import moe_permute_with_scale
```
Comment (huangzhilin-hzl) [inline test/manual/layers/moe/test_humming_fp8_dispatch.py]: This import uses the old module path and fails during test collection. Please update it to sglang.kernels.ops.moe.ep_moe_kernels.

Comment (guzekai01) [reply test/manual/layers/moe/test_humming_fp8_dispatch.py]: Thanks, fixed.

File: python/sglang/kernels/ops/moe/ep_moe_kernels.py [hunk 2]
```diff
@@ -1948,9 +1938,56 @@ def moe_permute(
...
+    num_experts: int,
+    use_int64_offset: bool = False,
+    is_ep: bool = False,
+    outputs: torch.Tensor | None = None,
+    scale_outputs: torch.Tensor | None = None,
+) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+    if inputs.size(0) != input_scale.size(0):
+        raise ValueError("inputs and input_scale must have the same number of rows")
+
+    outputs, src2dst, expert_offsets = moe_permute(
```
Comment (huangzhilin-hzl) [inline python/sglang/kernels/ops/moe/ep_moe_kernels.py:2086]: Non-blocking suggestion: could we fuse the two row-permutation launches in `moe_permute_with_scale`?

Currently, `moe_permute()` launches `deepep_permute_triton_kernel` for `hidden_states`, then `_moe_permute_rows()` launches the same kernel again for `input_scale`, reusing the same `src2dst`. A specialized fused kernel could copy both tensors using the same `dst_idx`, reducing one kernel launch and duplicate reads of `src2dst`/`topk_ids`. `ep_scatter` already follows a similar fused-copy approach.

File: python/sglang/srt/layers/moe/moe_runner/humming.py [hunk 3]
```diff
@@ -396,6 +405,63 @@ def apply_activation(self, inputs: torch.Tensor, outputs: torch.Tensor):
...
+        self.apply_activation(
+            inputs=buffers["gate_up_output"],
+            outputs=buffers["activation_output"],
+        )
+        return HummingMethod.may_quant_input(
+            layer=self.layer,
+            inputs=buffers["activation_output"],
+            quanted_input=buffers.get("quanted_down_input"),
+            sublayer_name="w2",
+        )
```
Comment (huangzhilin-hzl) [inline python/sglang/srt/layers/moe/moe_runner/humming.py:464]: One point that may be worth noting is the fallback behavior when the fused JIT activation/quantization conditions are not satisfied—for example, when W2 does not use per-token-per-group FP8 quantization with group size 128.

In that case, the code falls back to separate `apply_activation()` and `HummingMethod.may_quant_input()` calls. Under DeepEP low-latency dispatch, these operations process the full static grouped buffer, including padded expert slots, without using `expert_num_tokens` to skip invalid rows. This can introduce significant overhead.

For example, in the DSv4 pro EP32/BS32 decode trace with 5 verify tokens per request:

- Actual MoE input tokens: `32 × 5 = 160`
- Actual routed token-expert pairs with `top_k=6`: `160 × 6 = 960`
- Static capacity per expert: `256 max dispatch tokens/source rank × 32 source ranks = 8192`
- Physical rows across 13 local experts: `13 × 8192 = 106,496`

The trace shows:
- `act_and_mul_kernel`: grid `106,496`, approximately `583 μs` per layer
- `_quant_tensor_kernel`: grid `319,488`, approximately `657 μs` per layer

By comparison, the fused varlen path `silu_and_mul_masked_post_quant` launches `8192 × top_k(6) = 49,152` blocks and takes approximately `97 μs` per layer.

Comment (guzekai01) [reply python/sglang/srt/layers/moe/moe_runner/humming.py:464]: Thanks for calling out the cost of the generic fallback. I tightened the fast-path guard after rebasing: it now checks the fused kernel launch contract directly—a hidden dimension divisible by 256, hidden_dim / 8 no greater than 1024 threads, at most 256 experts, and enough threads to scan all expert counts—and no longer imposes unrelated divisibility-by-four constraints on padded tokens or FP32 scale groups.

For the Humming + DeepEP FP8 configuration in this PR, layer setup explicitly assigns W2 an FP8 group-128 input schema, so the intended path satisfies this specialized kernel contract. The manual case with max_tokens=63 and six group-128 scale groups covers a shape that the old guards rejected unnecessarily.

This does not remove the fallback for schemas the specialized kernel does not support, including non-group-128 W2 input quantization. Those configurations still process the padded buffer with separate activation and quantization kernels; supporting them efficiently requires a separate kernel-generalization follow-up. I am therefore leaving this thread open as a non-blocking optimization rather than claiming that the generic fallback has been eliminated.

File: python/sglang/kernels/ops/moe/ep_moe_kernels.py (hunk 2 above)
Comment (guzekai01) [reply python/sglang/kernels/ops/moe/ep_moe_kernels.py:2086]: Agreed that fusing the hidden-state and scale copies can save one launch and duplicate index reads. I kept it out of this PR because it is non-blocking and would expand the kernel correctness surface while this change is focused on enabling FP8 DeepEP dispatch. I will handle it as a follow-up optimization; the existing scale permutation remains functionally correct.

File: python/sglang/srt/layers/quantization/humming_utils.py [hunk 4]
```diff
@@ -35,6 +37,50 @@ def humming_is_layer_skipped(config: dict[str, Any], prefix: str):
...
+    dispatcher.set_quant_config(quant_config)
+
+
+def configure_humming_deepep_dispatch(layer: torch.nn.Module) -> bool:
+    if not get_moe_a2a_backend().is_deepep():
+        layer._humming_uses_deepep_fp8_dispatch = False
+        _set_humming_dispatcher_output_dtype(layer, "bf16")
+        return False
+
+    output_dtype = get_server_args().deepep_dispatcher_output_dtype
```
Comment (BBuf) [inline python/sglang/srt/layers/quantization/humming_utils.py]: Could we read this from `get_exec().moe.deepep_dispatcher_output_dtype`? Runtime code should use the resolved config bag; `get_server_args()` is the pristine startup copy, so runtime overrides can diverge here.

File: python/sglang/kernels/ops/moe/ep_moe_kernels.py [hunk 5]
```diff
@@ -2038,9 +2028,56 @@ def moe_permute(
...
+        num_experts=num_experts,
+        use_int64_offset=use_int64_offset,
+        is_ep=is_ep,
+    )
+    outputs = _moe_permute_rows(inputs, topk_ids, src2dst, outputs)
+
     return outputs, src2dst, expert_offsets
 
 
+def moe_permute_with_scale(
```
Comment (BBuf) [inline python/sglang/kernels/ops/moe/ep_moe_kernels.py:2073]: Could we move the scale-permutation and dispatcher-validation tests from #33884 into this PR? These helpers should be covered before this PR can land independently.

File: python/sglang/srt/layers/quantization/humming_utils.py (hunk 4 above)
Comment (guzekai01) [reply python/sglang/srt/layers/quantization/humming_utils.py]: Done

File: python/sglang/kernels/ops/moe/ep_moe_kernels.py (hunk 5 above)
Comment (guzekai01) [reply python/sglang/kernels/ops/moe/ep_moe_kernels.py:2073]: Done

**Review by BBuf — APPROVED** [review]

**Comment by BBuf** [comment]
图片
